promotion_fund/src/daemon.rs

```diff
@@ -0,0 +1,168 @@
+        )
+    }
+}
+
+impl Daemon {
+    pub fn new(
+        s3_current: S3Value,
+        solana_client: SolanaRpc,
+        file_sink: FileSinkClient<ServiceProviderPromotionFundV1>,
+        solana_check_interval: Option<Interval>,
+    ) -> Self {
+        Self {
+            s3_current,
+            solana_client,
+            file_sink,
+            solana_check_interval: solana_check_interval.unwrap_or(time::interval(Duration::MAX)),
+        }
+    }
+
+    pub async fn from_settings(
+        settings: &Settings,
+        file_sink: FileSinkClient<ServiceProviderPromotionFundV1>,
+    ) -> anyhow::Result<Self> {
+        let s3_current = fetch_s3_bps(&settings.file_store_output).await?;
+        let solana_client = SolanaRpc::new(&settings.solana).context("making solana client")?;
+        let check_timer = tokio::time::interval(settings.solana_check_interval);
+
+        Ok(Self::new(
+            s3_current,
+            solana_client,
+            file_sink,
+            Some(check_timer),
+        ))
+    }
+
+    pub async fn run(mut self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        loop {
+            tokio::select! {
+                _ = shutdown.clone() => break,
+                _ = self.solana_check_interval.tick() => self.handle_tick().await?
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn handle_tick(&mut self) -> Result<()> {
+        let solana_current = match fetch_solana_bps(&self.solana_client).await {
+            Ok(solana_current) => {
+                metrics::gauge!(PROMOTION_FUND_LAST_SOLANA_FETCH_TIME)
+                    .set(Utc::now().timestamp() as f64);
+                solana_current
+            }
+            Err(err) => {
+                tracing::error!(?err, "failed to get bps from solana");
+                return Ok(());
+            }
+        };
+
+        let action = compare_s3_and_solana_values(&self.s3_current, &solana_current);
+        match action {
+            Action::Noop => tracing::info!("nothing to do"),
+            Action::Write => {
+                tracing::info!(items = solana_current.len(), "writing new file");
+                self.store_solana_values(&solana_current);
+                write_protos(&self.file_sink, solana_current).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn store_solana_values(&mut self, promo_funds: &[ServiceProviderPromotionFundV1]) {
+        self.s3_current.clear();
+
+        for promo_fund_v1 in promo_funds {
+            self.s3_current
+                .insert(promo_fund_v1.service_provider, promo_fund_v1.bps);
+        }
+    }
+}
+
+pub async fn fetch_s3_bps(settings: &file_store::Settings) -> anyhow::Result<S3Value> {
+    let file_store = FileStore::from_settings(settings).await?;
+    let mut results = HashMap::new();
+
+    let all = file_store
+        .list_all(FileType::ServiceProviderPromotionFund.to_str(), None, None)
```

Comment:
anyway to have a date in this statement, so that over time we are pulling a list of thousands of files to just get the last one.  Maybe we could write something to disk?  and if it is not available on disk, we fall ack to just get everything?

Reply:
A `lookback_start_after` setting was added. It defaults to the unix epoch, but can be updated to a value closer to a recent file when the s3 bucket starts to fill up.